x-pack/plugins/security_solution/server/lib/telemetry/sender.ts

```diff
@@ -0,0 +1,371 @@
+    } else {
+      this.queue.push(...this.processEvents(events));
+    }
+  }
+
+  public processEvents(events: TelemetryEvent[]): TelemetryEvent[] {
+    return events.map(function (obj: TelemetryEvent): TelemetryEvent {
+      return copyAllowlistedFields(allowlistEventFields, obj);
+    });
+  }
+
+  private async sendIfDue() {
+    // this.logger.debug(`Send if due`);
+    if (this.isSending) {
+      return;
+    }
+
+    if (this.queue.length === 0) {
+      return;
+    }
+
+    try {
+      this.isSending = true;
+
+      // Checking opt-in status is relatively expensive (calls a saved-object), so
+      // we only check it when we have things to send.
+      this.isOptedIn = await this.telemetryStart?.getIsOptedIn();
+      if (!this.isOptedIn) {
+        this.logger.debug(`Telemetry is not opted-in.`);
+        this.queue = [];
+        this.isSending = false;
+        return;
+      }
+
+      const telemetryUrl = await this.fetchTelemetryUrl();
+      this.logger.debug(`Telemetry URL: ${telemetryUrl}`);
+
+      const clusterInfo = await this.fetchClusterInfo();
+      this.logger.debug(
+        `cluster_uuid: ${clusterInfo?.cluster_uuid} cluster_name: ${clusterInfo?.cluster_name}`
+      );
+
+      const licenseInfo = await this.fetchLicenseInfo();
+
+      const toSend: TelemetryEvent[] = cloneDeep(this.queue);
+      this.queue = [];
+
+      toSend.forEach((event) => {
+        event.cluster_uuid = clusterInfo.cluster_uuid;
+        event.cluster_name = clusterInfo.cluster_name;
+        this.copyLicenseFields(event, licenseInfo);
+      });
+
+      await this.sendEvents(toSend, telemetryUrl, clusterInfo.cluster_uuid);
+    } catch (err) {
+      this.logger.warn(`Error sending telemetry events data: ${err}`);
+      // throw err;
+      this.queue = [];
+    }
+    this.isSending = false;
+  }
+
+  private async fetchClusterInfo(): Promise<ESClusterInfo> {
+    if (!this.core) {
+      throw Error("Couldn't fetch cluster info because core is not available");
+    }
+    const callCluster = this.core.elasticsearch.legacy.client.callAsInternalUser;
+    return getClusterInfo(callCluster);
+  }
+
+  private async fetchTelemetryUrl(): Promise<string> {
+    const telemetryUrl = await this.telemetrySetup?.getTelemetryUrl();
+    if (!telemetryUrl) {
+      throw Error("Couldn't get telemetry URL");
+    }
+    return getV3UrlFromV2(telemetryUrl.toString(), 'alerts-endpoint'); // TODO: update
+  }
+
+  private async fetchLicenseInfo(): Promise<ESLicense | undefined> {
+    if (!this.core) {
+      return undefined;
+    }
+    try {
+      const callCluster = this.core.elasticsearch.legacy.client.callAsInternalUser;
+      const ret = await getLicense(callCluster, true);
+      return ret.license;
+    } catch (err) {
+      this.logger.warn(`Error retrieving license: ${err}`);
+      return undefined;
+    }
+  }
+
+  private copyLicenseFields(event: TelemetryEvent, lic: ESLicense | undefined) {
```

Comment:
nit -> I am suggesting that because I do not like updating object by their reference
```
private copyLicenseFields(lic: ESLicense | undefined) {
    if (lic) {
      return {
        uid: lic.uid,
        status: lic.status,
        type: lic.type,
        ...(lic.issued_to ? { issued_to: lic.issued_to } : {}),
        ...(lic.issuer ? { issuer: lic.issuer } : {}),
      };
    }
    return {};
  }
```